apm-agent-plugins/apm-scala-concurrent-plugin/src/test/scala/co/elastic/apm/agent/scala/concurrent/FutureInstrumentation2Spec.scala

```diff
@@ -0,0 +1,66 @@
+package co.elastic.apm.agent.scala.concurrent
+
+import java.util.concurrent.Executors
+
+import co.elastic.apm.agent.MockReporter
+import co.elastic.apm.agent.bci.ElasticApmAgent
+import co.elastic.apm.agent.configuration.{CoreConfiguration, SpyConfiguration}
+import co.elastic.apm.agent.impl.transaction.Transaction
+import co.elastic.apm.agent.impl.{ElasticApmTracer, ElasticApmTracerBuilder}
+import munit.FunSuite
+import net.bytebuddy.agent.ByteBuddyAgent
+import org.stagemonitor.configuration.ConfigurationRegistry
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, ExecutionContext, ExecutionContextExecutor, Future}
+
+class FutureInstrumentation2Spec extends FunSuite {
+
+  private var reporter: MockReporter = _
+  private var tracer: ElasticApmTracer = _
+  private var coreConfiguration: CoreConfiguration = _
+
+  override def beforeEach(context: BeforeEach): Unit = {
+    reporter = new MockReporter
+    val config: ConfigurationRegistry = SpyConfiguration.createSpyConfig
+    coreConfiguration = config.getConfig(classOf[CoreConfiguration])
+    tracer = new ElasticApmTracerBuilder().configurationRegistry(config).reporter(reporter).build
+    ElasticApmAgent.initInstrumentation(tracer, ByteBuddyAgent.install)
+    tracer.start()
+  }
+
+  override def afterEach(context: AfterEach): Unit = ElasticApmAgent.reset()
+
+  test("Scala Future should not propagate the tracing-context to unrelated threads") {
+    implicit val executionContext: ExecutionContextExecutor =
+      ExecutionContext.fromExecutor(Executors.newFixedThreadPool(10))
+
+    def startFuture(num: Int): Future[(Transaction, Int)] = {
+      Future {
+        println(s"thread=${Thread.currentThread().getId}, trace=${tracer.currentTransaction()} before num=$num")
+        val transaction = tracer.startRootTransaction(getClass.getClassLoader).withName("Transaction" + num).activate()
```

Comment:
This transaction is never deactivated on the thread and can therefore leak into other operations.

A more sensible test scenario would be to start a transaction before doing the multi-threaded operations and check if the transaction successfully propagates to the futures executed in different threads. I've tried this out and it even works when commenting out the instrumentations in `apm-agent-plugins/apm-scala-concurrent-plugin/src/main/resources/META-INF/services/co.elastic.apm.agent.bci.ElasticApmInstrumentation`. That's because Scala relies on the Java concurrent framework that we already instrument.

The instrumentations in this module are only needed when the context gets lost along the way which seems to be the case for some AsyncHttpClient callbacks.

```scala
val transaction = tracer.startRootTransaction(getClass.getClassLoader).withName("Transaction").activate()
val futures = (1 to 10).map(x => Future {
  Thread.sleep(10)
  println(s"thread=${Thread.currentThread().getId}, trace=${tracer.currentTransaction()}")
  val x = transaction == tracer.currentTransaction()
  assertEquals(transaction, tracer.currentTransaction())
  (transaction, x)
})
val future = Future.sequence(futures)
val result = Await.result(future, 10.seconds).toList
transaction.deactivate().end()
assertEquals(result.forall(x => x._2), true)
```

Reply:
> This transaction is never deactivated on the thread and can therefore leak into other operations.
But if I deactivate it, it won't propagate... right?

How do I "activate" the transaction inside the future, so that it's passed on to the next .map() operation, but while deactivating it so that it's not passed on to unrelated operation that happens on the same thread?

Looks like your example is using a single transaction. What I'm trying to simulate is a "http server" scenario, which has multiple transactions corresponding to http requests concurrently. If I take your example and make it run that 10 times in different threads I get the same issue.
https://gist.github.com/henrikno/71e54b7c2d43e63936fd33fdd45411bb
Interestingly it happens less often with ForkJoinPool than with FixedThreadPool.

If the test is wrong, could you help explain how it _should_ be used with multiple "requests"/threads handing different transactions concurrently (each of which will typically spawn nested futures to do the actual work)?

Reply:
The test in the linked gist looks good. I had to increase the number of threads though. Seems like it does expose some edge cases where the context isn't properly transferred.

To make investigation easier, we'd ideally want a test that's a bit simpler but that still fails.

Reply:
Ill spend some time on this tomorrow 🙂

Reply:
@felixbarny I am not sure how the elastic-agent handles the following case for JavaFutures:

1. Create multiple transactions (could be on different or the same threads) - https://gist.github.com/henrikno/71e54b7c2d43e63936fd33fdd45411bb#file-test-scala-L41
2. After you have multiple transactions active, create a new `Future` - https://gist.github.com/henrikno/71e54b7c2d43e63936fd33fdd45411bb#file-test-scala-L43

In this case, the problem seems to be, how does the `Future` from (2) know which transaction to use? 

That's how I currently understand the issue show-cased on the Gist.

Reply:
There can only be one transaction active on a thread at a time. I think that's maintained as each Future is either executed in its own thread concurrently or steals time from the main thread and executes synchronously.

> how does the Future from (2) know which transaction to use?

The [`map`](https://gist.github.com/henrikno/71e54b7c2d43e63936fd33fdd45411bb#file-test-scala-L43) schedules the `Future` for execution (by calling eventually `Executor#execute`). Whichever transaction is active on the current thread when the `Future` is scheduled is going to be restored when the `Future` is executed.

Reply:
@felixbarny That `map` is not actually being called on any `Future`. Therefore, it will actually create a new `Future`  and use whatever `Transaction` is active at that point. 

I think that is the exact problem here. Transactions are started in parallel but are not followed up on correctly to pass the correct context.

So somehow, we need to have Transactions created inside of a Future, actually be linked to that Future and continue the logic with `map`, `flatMap` on that Future.

I'm not sure if the above not already happens though, say: `Future(startTransaction()).flatMap(logic)` maybe always runs on the same thread and use the same transaction. But Im not sure about that.

Reply:
In my tests, I have commented out the scala-specific `Future` instrumentation. The tests do fail more often when doing that but when using a `ForkJoinPool`, it works a bit more reliably.

Calling `map` ends up calling `Executor#execute` and executing a `Future` ends up calling `Runnable#run`. These methods are instrumented by the java-concurrent plugin which captures and restores the context. Something must be going wrong in some cases there.

> So somehow, we need to have Transactions created inside of a Future, actually be linked to that Future and continue the logic with map, flatMap on that Future.

When transactions are activated, they are attached to the current thread. If on that thread, a future gets scheduled for execution via `map`, the transaction gets associated with that future and will be restored when the future is executed.

Reply:
@felixbarny So since the Future that creates the transaction will be the same Runnable that executes the underlying Futures using that transaction, the context should be passed along correctly?

Thanks for the information as always :)